src/librustdoc/clean/simplify.rs

```diff
@@ -73,6 +73,7 @@ crate fn where_clauses(cx: &DocContext<'_>, clauses: Vec<WP>) -> Vec<WP> {
     clauses.extend(
         lifetimes.into_iter().map(|(lt, bounds)| WP::RegionPredicate { lifetime: lt, bounds }),
     );
+    clauses.extend(equalities.map(|(lhs, rhs)| WP::EqPredicate { lhs, rhs }));
```

Comment:
This doesn't seem right; equalities should be after all parameters, not in between lifetimes and generics.

Reply:
In that case, I'll revert the second commit. I originally had an alternate approach:
> ...the extend statement for `params` could use `.iter()` instead and copy/clone where necessary

but I just realized that `params` is borrowed mutably in the `retain` (`filter`, in this PR) call, so it wouldn't work. It seems like I'll have to close this PR.

Reply:
I'm also not quite sure what the purpose of this PR is. Could you elaborate what you mean by "implicit `collect`"?

Reply:
I was under the impression that the `retain` call implicitly iterates through the collection, filters, and `collect`s the elements that are supposed to remain. (I remember seeing https://github.com/rust-lang/rust/issues/91497, though I didn't notice the subsequent PR to fix it, so `retain` seemed even less desirable.) In any case, I had hoped to avoid anything beyond the necessary costs of iterating and filtering.

Reply:
My understanding of `retain` is that it just moves memory around; there shouldn't be any implicit `collect`s. However, this rustdoc code does iterate twice.

This code could probably be cleaned up in a way that would improve the performance, but it'd likely be a larger change than just this. So I'm going to close this PR for now. Please feel free to re-open it (or open a new PR) if you find a way to get it working.